// appdynamics-lambda-tracer/config/agentConfigResponse.js
function AgentConfigResponse(response, agent) {
  this.config = response ? response : {};
  this.updatedAt = response ? Date.now() : undefined;
  this.agent = agent;
}

AgentConfigResponse.prototype.getLastUpdateTime = function () {
  return this.updatedAt;
}

AgentConfigResponse.prototype.update = function (configUpdate) {
  for (var prop in configUpdate) {
    this.config[prop] = configUpdate[prop];
  }
  this.updatedAt = Date.now();
}

AgentConfigResponse.prototype.getRegisteredBTs = function () {
  return this.config['bts'];
}

AgentConfigResponse.prototype.getRegisteredBackendInfo = function () {
  return this.config['backends'];
}

AgentConfigResponse.prototype.getApplicationId = function () {
  return this.config['application_id'];
}

AgentConfigResponse.prototype.getAccountGuid = function () {
  return this.config['account_guid'];
}

AgentConfigResponse.prototype.getTierId = function () {
  return this.config['tier_id'];
}

AgentConfigResponse.prototype.getControllerGuid = function () {
  return this.config['controller_guid'];
}

AgentConfigResponse.prototype.getTimeSkew = function () {
  if (this.config['time_skew'])
    return this.config['time_skew'];
  else
    return 0;
}

AgentConfigResponse.prototype.getGlobalAccountName = function () {
  return this.config['global_account_name'];
}

AgentConfigResponse.prototype.isLagging = function () {
  return this.config['lagging'];
}

module.exports.AgentConfigResponse = AgentConfigResponse;
